intercom-common/src/foreign_ty.rs

```diff
@@ -1,59 +1,174 @@
+extern crate std;
+
 use model::ComCrate;
 use syn;
 
+/// Detailed information of a Rust type.
+pub enum RustType<'e>
+{
+    Ident( &'e syn::Ident ),
+    Void,
+}
+
+/// Defines how a value is passed to/from Rust function.
+#[derive(PartialEq, PartialOrd, Clone)]
+pub enum PassBy
+{
+    Value,
+
+    Reference,
+
+    Ptr,
+}
+
+/// Details of a Rust type. Used for translation to other programming languages.
+pub struct TypeInfo<'s>
+{
+    /// The crate this type is associated with.
+    pub krate: &'s ComCrate,
+
+    /// The type in rust.
+    pub rust_type: RustType<'s>,
+
+    /// Specifies how the value is passed to/from a function. Default: PassBy::Value
+    pub pass_by: PassBy,
+
+    /// Is the Rust type defined as mutable? Default: false
+    pub is_mutable: bool,
+
+    /// The length of the array if the Rust type denotes an array.
+    pub array_length: Option<&'s syn::Expr>
+}
+
 pub trait ForeignTypeHandler
 {
     /// Gets the name for the 'ty'.
     fn get_name( &self, krate : &ComCrate, ty : &syn::Ident ) -> String;
 
     /// Gets the COM type for a Rust type.
-    fn get_ty( &self, krate : &ComCrate, ty : &syn::Type ) -> Option< String >;
+    fn get_ty<'a, 'b: 'a>( &self, krate : &'b ComCrate, ty : &'b syn::Type ) -> Option< TypeInfo<'a> >;
```

Comment:
And yes, this was a question. I'm not certain myself. My intuition says they are equal and the single lifetime would be more idiomatic (since that's what the borrowck more or less has to assume anyway), but I'm not sure if there is some difference.

Reply:
> What's the difference with the current code versus:

My version states that the parameters can live longer than the TypeInfo struct. That was rather important bit...

Reply:
> Currently for example struct CTypeHandler is defined somewhere above, then the > ForeignTypeHandler for it is impl'd after struct TypeInfoResolver and finally the impl CTypeHandler is a the very end of the file.

I'm not sure about that. In general I want a quick access to the data / data types defined in a module (or class in C++). But I really haven't made up my mind which is better, yet.

Reply:
Isn't that always the case with lifetime bounds on parameters though?

```rust
fn foo<'a>( a: &'a A ) -> &'a B
```

I always read that as:
> There is a lifetime `'a` which is constrained by the return value. The parameter `a` has to fulfill it.

The function could be called with a `'static` reference just fine for example.

In a similar way as:
```rust
fn foo( a: &SomeTrait )
```
> The parameter must fulfill `SomeTrait`, it doesn't need to be exactly `SomeTrait`.

Reply:
Unfortunately I don't remember the exact error I was getting from the borrows checker anymore. But it couldn't handle the situation without this additional information.

Reply:
But if I remember correctly part of the problem was that the TypeInfoResolver died before the parameters and they had the same lifetime.

The chapter on [Advanced Lifetimes](https://doc.rust-lang.org/book/second-edition/ch19-02-advanced-lifetimes.html) might help.

Reply:
> part of the problem

So having just `'a` didn't work? The `'b : 'a` was needed to get it compiling? This is good enough for me.

Reply:
Do you want that I change the order? 

I would prefer that we wait and see how this compares to the other functions for a while.

Reply:
Yeah, it's fine for now. We can throw things around later if we want. :)